### PR TITLE
Impl `C` and `swift(iOS)` bindings for `ergo-lib`

### LINKED ISSUE #47: [5000 SigmaUSD] C + Swift(iOS) bindings
## Scope
Implement an API feature-wise on par with Wasm bindings (ergo-lib-wasm). Basically an API that a wallet would need to be able to send and receive coins and tokens including input box selection, output box creation and transaction building and signing.

## Architecture
The core part is the common/shared source code with JNI (ergo-lib-jni) that should be implemented in ergo-lib-c-core crate.
The generic C bindings should be implemented in ergo-lib-c which should be used in Swift wrapper that should be implemented in ergo-lib-ios.

### C
The main idea is to pass Rust types wrapped in opaque pointers and expose functions to manipulate them on C side.
The crucial part is to free resources on the same side(usually Rust) where they were allocated. For this we need to expose a separate function for every "resource-producing" function that frees a resource passed from C. In case of Address see address_from_testnet and address_delete.

Use the already implemented Address type as an example. 

### Swift
Wraps functions from ergo-lib-c in classes for resource management and exposes methods converting to/from Swift types where necessary.

## Types
Using Wasm bindings as a reference we need the following types(with methods) implemented. For implmentation details see Wasm types and their methods in ergo-lib-wasm. Some method names are adopted for C types. See implemented Address type

Wallet:
from_secrets
sign_transaction

SecretKey:
dlog_from_bytes
to_bytes

ErgoStateContext:
new(from PreHeader)

PreHeader:
from_block_header

BlockHeader:
from_json

BlockHeaders:
from_json

UnsignedTransaction:
id
inputs
data_inputs
output_candidates
to_json
from_json

ErgoBox:
new
box_id
creation_height
tokens
ergo_tree
value
register_value
to_json
from_json

ErgoBoxCandidate:
creation_height
tokens
ergo_tree
value
register_value

Token:
new
id
amount

TokenId:
from_box_id
from_str
to_str

TokenAmount:
from_int64_t
as_int64_t

BoxValue:
from_int64_t
as_int64_t

Transaction:
id
to_json
from_json
inputs
data_inputs
output_candidates
outputs

DataInput:
new
box_id

TxBuilder:
new
set_data_inputs
build
box_selection
data_inputs
output_candidates
current_height
fee_amount
change_address
min_change_value

ErgoBoxCandidateBuilder:
new
set_min_box_value_per_byte
min_box_value_per_byte
set_value
value
calc_box_size_bytes
calc_min_box_value
set_register_value
register_value
delete_register_value
mint_token
add_token
build

BoxSelection:
new
boxes
change

SimpleBoxSelector:
select

ErgoBoxAssetsData:
new
value
tokens

## Collections
The following types represent a simple wrapper for Vec<T>. Feel free to explore the idea to use the single type (pointer+size) since Rust arrays and slices have the same representation in C (arrays).

ErgoBoxCandidates; 
Tokens;
DataInputs;
ErgoBoxAssetsDataList;
SecretKeys;
ErgoBoxes (instead of UnspentBoxes, OutputBoxes, DataInputBoxes)

All of them should implement the following methods:
new
len
get
add

The ErgoBoxes type should implement from_boxes_json method (see Wasm).

### PR DESCRIPTION
**Pull request by kettlebell**
Closes #47

### CODE REVIEW

File: bindings/ergo-lib-c/src/macros.rs [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+///#[no_mangle]
+///pub unsafe extern "C" fn ergo_wallet_block_headers_add(
+///    element_ptr: ConstBlockHeaderPtr,
+///    collection_ptr_out: BlockHeadersPtr,
+///) -> ErrorPtr {
+///    Error::c_api_from(collection_add(collection_ptr_out, element_ptr))
+///}
+///```
+#[macro_export]
+macro_rules! make_collection {
```
Comment (greenhat) [inline bindings/ergo-lib-c/src/macros.rs:67]: Neat!

File: bindings/ergo-lib-c/src/lib.rs [hunk 2]
```diff
@@ -70,11 +76,127 @@ pub unsafe extern "C" fn ergo_wallet_address_from_testnet(
...
+    }
+}
+
 #[no_mangle]
 pub extern "C" fn ergo_wallet_address_delete(address: AddressPtr) {
     address_delete(address)
 }
 
+// -------------------------------------------------------------------------------------------------
+// BlockHeader functions ---------------------------------------------------------------------------
```
Comment (greenhat) [inline bindings/ergo-lib-c/src/lib.rs:137]: Is it possible to use modules here? So that we don't end up with one huge lib.rs file.

**Comment by greenhat** [comment]
I was curious to see how is it going. Great job! I understand that it's WIP so ignore my comment if it is ahead of time. :)

**Comment by greenhat** [comment]
Please ignore the failing CI. I believe it's a flaky test and I made #474  for it.

**Comment by kettlebell** [comment]
> I was curious to see how is it going. Great job! I understand that it's WIP so ignore my comment if it is ahead of time. :)

I appreciate your feedback, thank you! I've been a bit time-constrained the past week but I'll keep plugging away.

File: bindings/ergo-lib-c/src/lib.rs (hunk 2 above)
Comment (kettlebell) [reply bindings/ergo-lib-c/src/lib.rs:137]: Yes for sure. I'll break it up.

File: bindings/ergo-lib-c/src/macros.rs (hunk 1 above)
Comment (kettlebell) [reply bindings/ergo-lib-c/src/macros.rs:67]: Thanks. I think it's worth the trouble of using nightly for `cbindgen` to generate this boilerplate.

Comment (greenhat) [reply bindings/ergo-lib-c/src/macros.rs:67]: Oh yeah,  I agree.

**Review by greenhat — APPROVED** [review]
Fantastic job! It's a huge piece of work and you nailed it! I appreciate the Swift test suite, comments, and extra steps you took to make the code more readable and maintainable! And I especially appreciate you have pushed forward through the mundane parts of this job!

I just have one "tiny" question. Is it still possible to rename the `ergo_wallet` prefix and header file to `ergo_lib` in `ergo-lib-c`? I hope a dumb grep search should do most of the job. If it's too much of a hassle let's leave it like this.
I'm sorry, I should've checked it in the beginning, but `ergo_wallet` prefix was based on the old crate's name before it was renamed to `ergo-lib`. :facepalm:

**Comment by kettlebell** [comment]
> Fantastic job! It's a huge piece of work and you nailed it! I appreciate the Swift test suite, comments, and extra steps you took to make the code more readable and maintainable! And I especially appreciate you have pushed forward through the mundane parts of this job!
> 
> I just have one "tiny" question. Is it still possible to rename the `ergo_wallet` prefix and header file to `ergo_lib` in `ergo-lib-c`? I hope a dumb grep search should do most of the job. If it's too much of a hassle let's leave it like this. I'm sorry, I should've checked it in the beginning, but `ergo_wallet` prefix was based on the old crate's name before it was renamed to `ergo-lib`. 🤦

Thanks a lot! Sure no problem, I'll  rename it.

**Comment by kettlebell** [comment]
Thank you also for the Github invite. In the future for situations like this, would you like me to merge the PR after approval and checks clear?

**Comment by greenhat** [comment]
> Thank you also for the Github invite. In the future for situations like this, would you like me to merge the PR after approval and checks clear?

Oh, it was long overdue. Yes, please.

**Pull request merged**